Automattic/wp-calypso
# Checkout: Add new PayPal payment method that uses PayPal JS

## Background

The PayPal Express NVP payment method system is being deprecated and we need to transition to using the PayPal JS library and the PPCP order system for our PayPal transactions. See https://github.com/Automattic/payments-shilling/issues/3039

## Proposed Changes

In this PR we add a new PayPal payment method to replace the old one. The new method uses PayPal's JS library to display the submit button and take appropriate actions.

> [!NOTE]
> This only changes the PayPal button in checkout. It does not yet change the button on the "Add new payment method"/"Change payment method" form. This is to reduce the number of changes in one PR.

> [!IMPORTANT]
> This does not enable the new payment method by default except in development. It leaves the PayPal Express payment method as-is until a the `checkout/paypal-ppcp` config flag is turned on.

Requires D166389-code

## Testing Instructions

- Apply D166389-code and point `public-api.wordpress.com` at your sandbox. You will also need to set some configuration on your sandbox; see that diff's testing instructions for more info.
- Make sure your currency is in USD.
- Use `calypso.localhost` on this branch when testing; calypso.live links will not work because the new payment method is disabled there.
- Add a product to your cart and visit checkout.
- Click "Edit" (if required) on the payment method step of checkout and select PayPal.
- Click to submit the transaction.
- Verify that you remain on checkout but see a new PayPal window pop up over checkout (this differs from PayPal Express which does a full-page redirect).
- Confirm the transaction inside the popup window (you will need to use a sandboxed PayPal account). The window should close when you're done.
- Verify that checkout redirects to the post-checkout page(s).
- Verify that the transaction is successful with no errors.

Also we need to make sure that if D166389-code is _not_ applied, this PR has no effect and PayPal continues to work.
